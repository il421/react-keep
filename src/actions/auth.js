import { firebase } from '../firebase/firebase';
import { toast } from 'react-toastify';

export const loading = (loading) => ({
  type: 'LOADING',
  loading,
});

export const login = (uid, name, url) => ({
  type: 'LOGIN',
  uid,
  name,
  url
});

export const startLogin = (email, password) => {
  return async (dispatch) => {
    try {
      dispatch(loading(true));
      await firebase.auth().signInWithEmailAndPassword(email, password);
      console.log('Logged in successfully');
    } catch (e) {
      console.log(e);
      toast.error(e.message);
      dispatch(loading(false));
    } finally {
      dispatch(loading(false));
    }
  };
};

export const startSignUp = (email, password) => {
  return async (dispatch) => {
    try {
      dispatch(loading(true));
      await firebase.auth().createUserWithEmailAndPassword(email, password);
      console.log('Logged in successfully');
    } catch (e) {
      console.log(e);
      toast.error(e.message);
      dispatch(loading(false));
    } finally {
      dispatch(loading(false));
    }
  };
};

export const logout = () => ({
  type: 'LOGOUT'
});

export const startLogout = () => {
  return async () => {
    try {
      await firebase.auth().signOut();
      console.log('Logged out successfully');
    } catch (e) {
      console.log(e);
      toast.error(e.message);
    }
  };
};
