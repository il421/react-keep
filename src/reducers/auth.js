export default (state = {}, action) => {
  switch (action.type) {
  case 'LOGIN':
    return {
      uid: action.uid,
      name: action.name,
      url: action.url
    };
  case 'LOGOUT':
    return {};
  default:
    return state;
  }
};
