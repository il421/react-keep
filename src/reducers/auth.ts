import {
  AuthActionsTypes,
  AuthStoreState,
  LoadingAction,
  LoginAction,
  LogoutAction,
  UserModalToggle,
} from "../store/store.types";

export type AuthAction =
  | LoginAction
  | LogoutAction
  | LoadingAction
  | UserModalToggle;

export const defaultAuthStore: AuthStoreState = {
  name: null,
  url: null,
  uid: "",
  email: "",
  loading: false,
  isUserModalOpen: false,
};

export default (
  state = defaultAuthStore,
  action: AuthAction
): AuthStoreState | object => {
  switch (action.type) {
    case AuthActionsTypes.login:
      return {
        ...state,
        uid: action.uid,
        name: action.name,
        url: action.url,
        email: action.email,
      } as AuthStoreState;

    case AuthActionsTypes.logout:
      return defaultAuthStore;

    case AuthActionsTypes.loading:
      return {
        ...state,
        loading: action.loading,
      } as AuthStoreState;

    case AuthActionsTypes.toggleModal:
      return {
        ...state,
        isUserModalOpen: action.isUserModalOpen,
      } as AuthStoreState;

    default:
      return state;
  }
};
