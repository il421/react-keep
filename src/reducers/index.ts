export * from "./auth";
export * from "./collaborators";
export * from "./notes";
export * from "./tags";
export * from "./filters";
