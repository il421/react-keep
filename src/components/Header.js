import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout } from '../actions/auth';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faBars, faSignOutAlt, faTimes } from '@fortawesome/free-solid-svg-icons'
import SideNav from 'react-simple-sidenav';

export const Header = ({ startLogout, auth }) => {
  let [ showNav, onSetSidebarOpen ] = useState(false);
  const sidenavStyles = {
    title: {
      backgroundColor: '#4abdac',
      padding: '0 3rem 0 0',
      textAlign: 'right',
    },
    nav: {
      backgroundColor: '#4abdac',
      width: '300px'
    }
  }

  return (
    <header className="header">
      <div className="content-container">
        <div className="header__content">
          <div className="header__sidebar sidebar">
            <button className="button--sidenav" onClick={() => onSetSidebarOpen(showNav = true)}>
              <FontAwesomeIcon icon={ faBars } size="2x" />
            </button>
            <SideNav
              showNav={ showNav }
              onHideNav={() => onSetSidebarOpen(showNav = false)}
              navStyle={ sidenavStyles.nav }
              titleStyle={ sidenavStyles.title }
              items={[]}
              title={
                <button className="button--sidenav" onClick={() => onSetSidebarOpen(showNav = false)}>
                  <FontAwesomeIcon icon={ faTimes } size="1x" />
                </button>
              }
            />
          </div>
          <Link to="/dashboard">
            <h1 className="header__title">Keep Me</h1>
          </Link>
          <div className="header__user-box user-box">
            <div className="user-box__photo">
              <img src={auth.url} width={35} height={35} />
            </div>
            <div className="user-box__name show-for-mobile">{ auth.name }</div>
            <button className="button button--link" onClick={ startLogout }>
              <FontAwesomeIcon icon={ faSignOutAlt } size="1x" />
            </button>
          </div>
        </div>
      </div>
    </header>
  );
}


const mapDispatchToProps = (dispatch) => ({
  startLogout: () => dispatch(startLogout())
});

const mapStateToProps = (state) => {
  return {
    auth: state.auth,
    url: state.url,
  }
};

export default connect(mapStateToProps, mapDispatchToProps)(Header);
