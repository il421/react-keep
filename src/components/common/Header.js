import React from 'react';
import { connect } from 'react-redux';

import { startLogout } from '../../actions/auth';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import Search from './Search';

export const Header = ({ startLogout, auth, showSidebar }) => {

  return (
    <header className="header" onClick={(evt) => evt.stopPropagation()}>
      <div className="content-container">
        <div className="header__content">
          <div className="header__sidebar">

            <button className="button button--sidenav" onClick={() => showSidebar()}>
              <FontAwesomeIcon icon="bars" size="2x" />
            </button>

          </div>
          <div>
            <h1 className="header__title">Keep Me</h1>
          </div>
          <div className="header__search">
            <Search />
          </div>
          <div className="header__user-box user-box">
            <div className="user-box__photo">
              <img src={ auth.url } width={ 50 } height={ 50 } />
            </div>
            <div className="user-box__name show-for-mobile">{ auth.name }</div>
            <button className="user-box__logout button button--sidenav" onClick={ startLogout }>
              <FontAwesomeIcon icon="sign-out-alt" size="lg" />
            </button>
          </div>
        </div>
      </div>
    </header>
  );
};


const mapDispatchToProps = (dispatch) => ({
  startLogout: () => dispatch(startLogout())
});

const mapStateToProps = (state) => {
  return {
    auth: state.auth,
    url: state.url,
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Header);
