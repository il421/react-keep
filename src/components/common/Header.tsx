import React from "react";
import { connect } from "react-redux";
import { useHistory } from "react-router-dom";
import { stringify } from "query-string";
import { AuthStoreState, Store } from "../../store/store.types";
import { QueryKeys, RouteActions } from "../../routers/Routing";
import { startLogout } from "../../actions/auth";
import Search from "./Search";
import { Controllers } from "./Controllers";
import {
  IconButton,
  UserPhoto,
  ContentContainer,
  LinkButton
} from "../ui-components";
import "../../styles/components/header/_header.scss";
import "../../styles/components/header/_user-box.scss";
import "../../styles/components/common/_visibility.scss";
import { ThunkDispatch } from "redux-thunk";

interface HeaderProps {
  showSidebar: (value: boolean) => void;
}

interface StateProps {
  auth: AuthStoreState;
}

interface DispatchProps {
  startLogout: () => void;
}

type Props = HeaderProps & DispatchProps & StateProps;

const Header: React.FunctionComponent<Props> = ({
  startLogout,
  auth,
  showSidebar
}) => {
  const history = useHistory();

  const onClickHandler = (key: keyof typeof QueryKeys) => {
    const query = stringify({
      [key]: RouteActions.edit
    });
    history.push(`${history.location.pathname}?${query}`);
  };

  return (
    <header className="header" onClick={evt => evt.stopPropagation()}>
      <ContentContainer>
        <div className="header__content">
          <IconButton
            className="header__bar"
            icon="bars"
            size="2x"
            onClick={() => showSidebar(true)}
          />

          <h1 className="header__title">Keep Me</h1>

          <Search wrapperClass="header__search" />

          <Controllers className="hide-for-mobile" />

          <div className="header__user-box user-box">
            {auth.name && auth.url && (
              <UserPhoto
                className="user-box__photo"
                src={auth.url}
                height={50}
                width={50}
              />
            )}
            <LinkButton
              text={auth.name ?? "Update profile"}
              type="button"
              className="user-box__link"
              onClick={() => onClickHandler(QueryKeys.user)}
            />
            <IconButton
              className="user-box__logout"
              icon="sign-out-alt"
              size="lg"
              onClick={startLogout}
            />
          </div>
        </div>
      </ContentContainer>
    </header>
  );
};

const mapDispatchToProps = (
  dispatch: ThunkDispatch<{}, {}, any>
): DispatchProps => ({
  startLogout: () => dispatch(startLogout())
});

const mapStateToProps = (state: Store): StateProps => {
  return {
    auth: state.auth
  };
};

export default connect<StateProps, DispatchProps, HeaderProps, Store>(
  mapStateToProps,
  mapDispatchToProps
)(Header);
