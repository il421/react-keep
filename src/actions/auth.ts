import {firebase} from "../firebase/firebase";
import {toast} from "react-toastify";
import {AuthActionsTypes, LoadingAction, LoginAction, LogoutAction, Store, UpdateUser} from "../store/store.types";
import {Dispatch} from "redux";
import { getMessage, Message } from "../common";

export const loading = (loading: boolean): LoadingAction => ({
  type: AuthActionsTypes.loading,
  loading,
});

export const login = (uid: string, name: string | null, url :string | null): LoginAction => ({
  type: AuthActionsTypes.login,
  uid,
  name,
  url
});

export const logout = (): LogoutAction => ({
  type: AuthActionsTypes.logout
});

export const startLogin = (email: string, password: string) => {
  return async (dispatch: Dispatch) => {
    try {
      dispatch(loading(true));
      await firebase.auth().signInWithEmailAndPassword(email, password);
      console.log(getMessage(Message.successLoggedIn));
    } catch (e) {
      console.log(e);
      toast.error(e.message);
      dispatch(loading(false));
    } finally {
      dispatch(loading(false));
    }
  };
};

export const startSignUp = (email: string, password: string) => {
  return async (dispatch: Dispatch) => {
    try {
      dispatch(loading(true));
      await firebase.auth().createUserWithEmailAndPassword(email, password);
      console.log(getMessage(Message.successLoggedIn));
    } catch (e) {
      console.log(e);
      toast.error(e.message);
      dispatch(loading(false));
    } finally {
      dispatch(loading(false));
    }
  };
};

export const startLogout = () => {
  return async () => {
    try {
      await firebase.auth().signOut();
      console.log(getMessage(Message.successLoggedOut));
    } catch (e) {
      console.log(e);
      toast.error(e.message);
    }
  };
};

export const updateUserData = (data: UpdateUser) => {
  return async (dispatch: Dispatch, getState: () => Store) => {
    try {
      dispatch(loading(true));
      const currentUser = await firebase.auth().currentUser;
      if (currentUser) {
        const auth = getState().auth;

        await currentUser.updateProfile({
          displayName: data.displayName,
          photoURL: data.photoURL
        });
        dispatch(login(auth.uid, data.displayName, auth.url));
      }

      toast.success(getMessage(Message.successUpdated));
      console.log(getMessage(Message.successUpdated));
    } catch (e) {
      console.log(e);
      toast.error(e.message);
    } finally {
      dispatch(loading(false));
    }
  };
};
