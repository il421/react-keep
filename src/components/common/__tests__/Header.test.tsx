import React from "react";
import { shallow, ShallowWrapper } from "enzyme";
import { Header, Props } from "../Header";
import { defaultAuthStore } from "../../../reducers/auth";

let props: Props, wrapper: ShallowWrapper<Props, any>;
const mockHistoryPush = jest.fn();

const history = {
  push: mockHistoryPush,
  location: {
    pathname: "pathname",
  },
};
beforeEach(() => {
  props = {
    showSidebar: jest.fn(),
    startLogout: jest.fn(),
    history: history as any,
    auth: defaultAuthStore,
  };

  wrapper = shallow<Props>(<Header {...props} />);
});

test("should render Header correctly", () => {
  expect(wrapper.debug()).toMatchSnapshot();
});

test("should show side bar if click on burger button", () => {
  wrapper.find("IconButton").at(0).prop<any>("onButtonClick")();
  expect(props.showSidebar).toHaveBeenLastCalledWith(true);
});

test("should open user edit modal by clicking on user name link", () => {
  wrapper.find("LinkButton").at(0).prop<any>("onClick")();
  expect(mockHistoryPush).toHaveBeenLastCalledWith("pathname?user=edit");
});

test("should open user edit modal by clicking on user cog icon", () => {
  wrapper.find("IconButton").at(1).prop<any>("onButtonClick")();
  expect(mockHistoryPush).toHaveBeenLastCalledWith("pathname?user=edit");
});

test("should logout if click on an logout button", () => {
  wrapper.find("IconButton").at(2).prop<any>("onButtonClick")();
  expect(props.startLogout).toHaveBeenLastCalledWith();
});
