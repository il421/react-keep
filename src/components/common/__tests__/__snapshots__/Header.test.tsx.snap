// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header correctly 1`] = `
"<header className=\\"header\\" onClick={[Function: onClick]}>
  <ContentContainer>
    <div className=\\"header__content\\">
      <IconButton className=\\"header__bar\\" icon=\\"bars\\" size=\\"2x\\" onButtonClick={[Function: onButtonClick]} />
      <h1 className=\\"header__title\\">
        KEEP ME
      </h1>
      <Connect(SearchBase) wrapperClass=\\"header__search\\" />
      <Controllers className=\\"hide-for-mobile\\" openDialog={[Function: openDialog]} />
      <div className=\\"header__user-box user-box\\">
        <LinkButton text=\\"Update profile\\" type=\\"button\\" className=\\"user-box__link\\" onClick={[Function: onClick]} />
        <IconButton className=\\"user-box__icon\\" icon=\\"users-cog\\" size=\\"lg\\" onButtonClick={[Function: onButtonClick]} />
        <IconButton className=\\"user-box__logout\\" icon=\\"sign-out-alt\\" size=\\"lg\\" onButtonClick={[Function: onButtonClick]} />
      </div>
    </div>
  </ContentContainer>
  <Modal isOpen={false} onRequestClose={[Function: onRequestClose]} className=\\"user-box__logout logout\\" ariaHideApp={false} portalClassName=\\"ReactModalPortal\\" bodyOpenClassName=\\"ReactModal__Body--open\\" role=\\"dialog\\" closeTimeoutMS={0} shouldFocusAfterRender={true} shouldCloseOnEsc={true} shouldCloseOnOverlayClick={true} shouldReturnFocusAfterClose={true} parentSelector={[Function: parentSelector]}>
    <ConfirmDialog title=\\"Do you want to logout?\\" handleConfirm={[Function: mockConstructor]} closeDialog={[Function: closeDialog]} className=\\"logout__dialog\\" buttonsProps={{...}} />
  </Modal>
</header>"
`;
